lib/src/sink_base.dart

```diff
@@ -0,0 +1,168 @@
+// Copyright (c) 2021, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'dart:async';
+import 'dart:convert';
+
+import 'package:meta/meta.dart';
+
+import 'async_memoizer.dart';
+
+/// An abstract class that implements [EventSink] in terms of [onData],
+/// [onError], and [onClose] methods.
+///
+/// This takes care of ensuring that events can't be added after [close] is
+/// called.
+abstract class EventSinkBase<T> implements EventSink<T> {
+  /// Whether [close] has been called and no more events may be written.
+  bool get _closed => _closeMemo.hasRun;
+
+  @override
+  void add(T data) {
+    _checkCanAddEvent();
+    onAdd(data);
+  }
+
+  /// A method that handles data events that are passed to the sink.
+  @visibleForOverriding
+  void onAdd(T data);
+
+  @override
+  void addError(Object error, [StackTrace? stackTrace]) {
+    _checkCanAddEvent();
+    onError(error, stackTrace);
+  }
+
+  /// A method that handles error events that are passed to the sink.
+  @visibleForOverriding
+  void onError(Object error, [StackTrace? stackTrace]);
+
+  @override
+  Future<void> close() => _closeMemo.runOnce(onClose);
```

Comment:
Having an `EventSink` return a `Future` is dangerous since the return type of `EventSink.close` is `void`. That means that *no code expecting an `EventSink` will handle the future*.
Returning a future here is misleading and likely to cause problems for users.